open-ce/build_tree.py

```diff
@@ -443,6 +449,59 @@ def get_external_dependencies(self, variant):
                                               variant["mpi_type"], variant["cudatoolkit"])
         return self._external_dependencies.get(variant_string, [])
 
+    def get_installable_packages(self, variant_str):
+        installable_packages =  set()
+
+        def check_matching(deps_set, dep_to_be_added):
+            # If exact match already present in the set, no need to add again
+            if dep_to_be_added in deps_set:
+                return None
+
+            # Check only dependency name if it is present
+            # For e.g. If dep_to_be_added is tensorflow-base >=2.4.* and set has tensorflow-base
+            dep_name_to_be_added = dep_to_be_added.split()[0]
+            if dep_name_to_be_added in deps_set and len(dep_to_be_added.split()) > 1:
+                deps_set.remove(dep_name_to_be_added)
+                return dep_to_be_added
+
+            # For e.g. If set has tensorflow-base 2.4.* and dep_to_be_added is
+            # either just tensorflow-base or tensorflow-base >=2.4.*
+            for dep in deps_set:
+                dep_name_from_set = dep.split()[0]
+                if dep_name_to_be_added == dep_name_from_set:
+                    return None
+
+            # If no match found, just add it
+            return dep_to_be_added
+
+        def _get_unique_deps_names(dependencies):
+            deps = set()
+            if not dependencies is None:
+                for dep in dependencies:
+                    generalized_dep = utils.generalize_version(dep)
+                    dep_to_update = check_matching(deps, generalized_dep)
+                    if dep_to_update:
+                        deps.add(dep_to_update)
+            return deps
+
+        def check_and_add(dependencies, parent_set):
+            dependencies = _get_unique_deps_names(dependencies)
+
+            for dep in dependencies:
+                pack_to_add = check_matching(parent_set, dep)
+                if pack_to_add:
+                    parent_set.add(pack_to_add)
+
+            return parent_set
+
+        for build_command in self.build_commands:
```

Comment:
I'm not sure if this will work. This will probably contain the dependencies for every variant that has been seen up to this point.

Reply:
You'll probably need to keep the CondaEnvFileGenerator call where it was previously, and still pass in the build_commands.

Reply:
I'll try that. But so far, I've tested TF/dali/lightGBM recipes and environment files created for them. They were all working. Do you think moving the CondaEnvFileGenerator call where it was will reduce the number of dependencies (duplicate ones)?

Reply:
I think now I understand what you're saying. I'll make the changes.